[Docs] Improve Bloom filter topic (#17547)
apache/druid: docs/development/extensions-core/bloom-filter.md

  -->

To use the Apache Druid&circledR; Bloom filter extension, include `druid-bloom-filter` in the extensions load list. See [Loading extensions](../../configuration/extensions.md#loading-extensions) for more information.

> This extension adds the ability to both construct Bloom filters from query results, and filter query results by testing

Reviewer: ```suggestion
This extension adds the abilities to construct Bloom filters from query results and to filter query results by testing
```

> against a Bloom filter. A Bloom filter is a probabilistic data structure for performing a set membership check. A Bloom

Reviewer: ```suggestion
against a Bloom filter. A Bloom filter is a probabilistic data structure to check for set membership. A Bloom
```

> filter is a good candidate to use with Druid for cases where an explicit filter is impossible, such as filtering a query

Reviewer: ```suggestion
filter is a good candidate to use when an explicit filter is impossible, such as filtering a query
```
Don't think we need "for Druid" since these are the Druid docs

against a set of millions of values.

Following are some characteristics of Bloom filters:

> - Bloom filters are highly space efficient compared to using a HashSet.

Reviewer: ```suggestion
- Bloom filters are significantly more space efficient than HashSets.
```

> - Because of the probabilistic nature of Bloom filters, false positive results are possible. For example, the `test()` function might return `true` for an element that wasn't inserted into the filter.

Reviewer: ```suggestion
- Because they are probabilistic, false positive results are possible with Bloom filters. For example, the `test()` function might return `true` for an element that not within the filter.
```

> - Because of the probabilistic nature of Bloom filters, false positive results are possible. For example, the `test()` function might return `true` for an element that wasn't inserted into the filter.

Author: ```suggestion
- Because they are probabilistic, false positive results are possible with Bloom filters. For example, the `test()` function might return `true` for an element that is not within the filter.
```

- False negatives are not possible. If an element is present, `test()` always returns `true`.
- The false positive probability of this implementation is fixed at 5%. Increasing the number of entries that the filter can hold can decrease this false positive rate in exchange for overall size.

> - Bloom filters are sensitive to the number of inserted elements. You must specify the expected number of entries when you create the Bloom filter. If the number of insertions exceeds the specified number of entries, the false positive probability increases accordingly.

Reviewer: ```suggestion
- Bloom filters are sensitive to the number of inserted elements. You must specify the expected number of entries at creation time. If the number of insertions exceeds the specified number of entries, the false positive probability increases accordingly.
```

This extension is based on `org.apache.hive.common.util.BloomKFilter` from `hive-storage-api`. Internally,
this implementation uses Murmur3 as the hash algorithm.

> The following example shows how to construct a BloomKFilter externally with Java to use as a filter in a Druid query:

Reviewer: ```suggestion
The following Java example shows how to construct a BloomKFilter externally:
```

String base64Serialized = Base64.encodeBase64String(byteArrayOutputStream.toByteArray());
```

> You can then use this string in the native or SQL Druid query.

Reviewer: ```suggestion
You can then use the Base64 encoded string in JSON-based or SQL-based queries in Druid.
```

}
```

|Property|Description|Required|
|--------|-----------|--------|

> |`type`|Filter type. Should always be `bloom`.|Yes|

Reviewer: ```suggestion
|`type`|Filter type. Set to `bloom`.|Yes|
```

|`dimension`|Dimension to filter over.|Yes|
|`bloomKFilter`|Base64 encoded binary representation of `org.apache.hive.common.util.BloomKFilter`.|Yes|
|`extractionFn`|[Extraction function](../../querying/dimensionspecs.md#extraction-functions) to apply to the dimension values.|No|

### Serialized format for BloomKFilter

Serialized BloomKFilter format:

- 1 byte for the number of hash functions.
- 1 big-endian integer for the number of longs in the bitset.
- Big-endian longs in the BloomKFilter bitset.

`org.apache.hive.common.util.BloomKFilter` provides a method to serialize Bloom filters to `outputStream`.

### Filter SQL queries

You can use Bloom filters in SQL `WHERE` clauses with the `bloom_filter_test` operator:

```sql
SELECT COUNT(*) FROM druid.foo WHERE bloom_filter_test(<expr>, '<serialized_bytes_for_BloomKFilter>')
```

### Expression and virtual column support

The Bloom filter extension also adds a Bloom filter [Druid expression](../../querying/math-expr.md) which shares syntax
with the SQL operator.

```sql
bloom_filter_test(<expr>, '<serialized_bytes_for_BloomKFilter>')
```

## Bloom filter query aggregator

> You can create an input for a `BloomKFilter` from a Druid query with the `bloom` aggregator. Make sure to set a reasonable value for the `maxNumEntries` parameter, which is the maximum number of distinct entries that the Bloom filter can represent without increasing the false positive rate. It may be worth performing a query using

Reviewer: ```suggestion
You can create an input for a `BloomKFilter` from a Druid query with the `bloom` aggregator. Make sure to set a reasonable value for the `maxNumEntries` parameter to specify the maximum number of distinct entries that the Bloom filter can represent without increasing the false positive rate. Try performing a query using
```

    }
```

|Property|Description|Required|
|--------|-----------|--------|

> |`type`|Aggregator type. Should always be `bloom`.|Yes|

Reviewer: ```suggestion
|`type`|Aggregator type. Set to `bloom`.|Yes|
```

}
```

Example response:

```json
[
  {
    "timestamp":"2015-09-12T00:00:00.000Z",
    "result":{"userBloom":"BAAAJhAAAA..."}
  }
]
```

> Ordering results by a Bloom filter aggregator, for example in a TopN query, can be resource-intensive. This is because the operation performs an expensive linear scan of the filter to approximate the count of items added to the set by counting the number of set bits. We recommend ordering by an alternative aggregation method.

Reviewer: ```suggestion
We recommend ordering by an alternative aggregation method instead of ordering results by a Bloom filter aggregator.
Ordering results by a Bloom filter aggregator can be resource-intensive because Druid performs an expensive linear scan of the filter to approximate the count of items added to the set by counting the number of set bits. 
```
Suggest putting the recommendation first. Then the explanation.
